db/kv_checksum.h

```diff
@@ -0,0 +1,307 @@
+  void UpdateT(uint64_t old_ts_checksum, uint64_t new_ts_checksum);
+
+  ProtectionInfo<N> info_;
+};
+
+template <size_t N>
+struct ProtectionInfoKVOTC {
+  ProtectionInfoKVOTC() {
+    static_assert(std::is_standard_layout<ProtectionInfoKVOTC<N>>::value, "");
+  }
+
+  ProtectionInfoKVOT<N> StripC(ColumnFamilyId column_family_id) const;
+
+  void UpdateK(uint64_t old_key_checksum, uint64_t new_key_checksum) {
+    kvot_.UpdateK(old_key_checksum, new_key_checksum);
+  }
+  void UpdateV(uint64_t old_value_checksum, uint64_t new_value_checksum) {
+    kvot_.UpdateV(old_value_checksum, new_value_checksum);
+  }
+  void UpdateO(ValueType old_op_type, ValueType new_op_type) {
+    kvot_.UpdateO(old_op_type, new_op_type);
+  }
+  void UpdateT(uint64_t old_ts_checksum, uint64_t new_ts_checksum) {
+    kvot_.UpdateT(old_ts_checksum, new_ts_checksum);
+  }
+  void UpdateC(ColumnFamilyId old_column_family_id,
+               ColumnFamilyId new_column_family_id);
+
+  ProtectionInfoKVOT<N> kvot_;
+};
+
+template <size_t N>
+struct ProtectionInfoKVOTS {
+  ProtectionInfoKVOTS() {
+    static_assert(std::is_standard_layout<ProtectionInfoKVOTS<N>>::value, "");
+  }
+
+  ProtectionInfoKVOT<N> StripS(SequenceNumber sequence_number) const;
+
+  void UpdateK(uint64_t old_key_checksum, uint64_t new_key_checksum) {
+    kvot_.UpdateK(old_key_checksum, new_key_checksum);
+  }
+  void UpdateV(uint64_t old_value_checksum, uint64_t new_value_checksum) {
+    kvot_.UpdateV(old_value_checksum, new_value_checksum);
+  }
+  void UpdateO(ValueType old_op_type, ValueType new_op_type) {
+    kvot_.UpdateO(old_op_type, new_op_type);
+  }
+  void UpdateT(uint64_t old_ts_checksum, uint64_t new_ts_checksum) {
+    kvot_.UpdateT(old_ts_checksum, new_ts_checksum);
+  }
+  void UpdateS(SequenceNumber old_sequence_number,
+               SequenceNumber new_sequence_number);
+
+  ProtectionInfoKVOT<N> kvot_;
+};
+
+// Non-persistent: this XOR-folding function produces different outcomes for
+// different endiannesses so its output must not be persisted.
+//
+// Only supported for `N` and `sizeof(T)` that are powers of two and that
+// cannot be larger than eight.
+template <size_t N, class T>
+void NPFoldXor(unsigned char* dst, T info) {
+  static_assert(sizeof(T) > 0 && sizeof(T) <= 8, "");
+  static_assert((sizeof(T) & (sizeof(T) - 1)) == 0, "");
+  static_assert(N > 0 && N <= 8, "");
+  static_assert((N & (N - 1)) == 0, "");
+
+  unsigned char* info_bytes = reinterpret_cast<unsigned char*>(&info);
+  uint64_t dst_val = 0;
+  memcpy(&dst_val, dst, N);
+  // In C++17 the below can be a static-if.
+  if (N < sizeof(T)) {
+    assert(sizeof(T) % N == 0);
+    for (size_t i = 0; i < sizeof(T); i += N) {
+      uint64_t info_val = 0;
+      memcpy(&info_val, info_bytes, N);
+      dst_val = dst_val ^ info_val;
+      info_bytes += N;
+    }
+  } else {
+    // TODO(ajkr): For calling `NPFoldXor()` multiple times with `info`s smaller
```

Comment:
That's a misleading anti-pattern to me. Although you can technically get away with one of the mix-ins not being hashed, your overall protection is really only as strong as your smallest hash being xored in.

You can rely on our stock hash functions to be lossless for (fixed) input sizes up to their output size, by the way. Ping me if you want a lossless hash construction for smaller sizes.